ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/validation/ErrorFactories.scala

```diff
@@ -26,6 +27,128 @@ import io.grpc.{Metadata, StatusRuntimeException}
+        errorCodesVersionSwitcher.choose(
+          v1 = {
+            val status = io.grpc.Status.ABORTED
+              .withDescription(s"Failure: ${t.getClass.getSimpleName}: ${t.getMessage}")
+              .withCause(t)
+            val statusBuilder = GrpcStatus.toJavaBuilder(status)
+            GrpcStatus.buildStatus(Map.empty, statusBuilder)
+          },
+          v2 = LedgerApiErrors.CommandErrors.CommandSubmissionFailure
+            .Reject(
+              messagePrefix = "Failure",
+              throwableO = Some(t),
+            )
+            .asGrpcStatusFromContext,
+        )
+      }
+
+      def ingressBufferFull()(implicit
+          contextualizedErrorLogger: ContextualizedErrorLogger
+      ): Status = {
+        errorCodesVersionSwitcher.choose(
+          v1 = {
+            val status = io.grpc.Status.RESOURCE_EXHAUSTED
+              .withDescription("Ingress buffer is full")
+            val statusBuilder = GrpcStatus.toJavaBuilder(status)
+            GrpcStatus.buildStatus(Map.empty, statusBuilder)
+          },
+          v2 = SubmissionErrors.ParticipantBackpressure
+            .Rejection("Ingress buffer is full")
+            .asGrpcStatusFromContext,
+        )
+      }
+
+      def queueClosed()(implicit
+          contextualizedErrorLogger: ContextualizedErrorLogger
+      ): Status = {
+        errorCodesVersionSwitcher.choose(
+          v1 = {
+            val status = io.grpc.Status.ABORTED.withDescription("Queue closed")
+            val statusBuilder = GrpcStatus.toJavaBuilder(status)
+            GrpcStatus.buildStatus(Map.empty, statusBuilder)
+          },
+          v2 = LedgerApiErrors.CommandErrors.CommandSubmissionFailure
+            .Reject(
+              messagePrefix = "Queue closed",
+              throwableO = None,
+            )
+            .asGrpcStatusFromContext,
+        )
+      }
+
+    }
+
+    object CompletionResponse {
+      def timeout()(implicit
+          contextualizedErrorLogger: ContextualizedErrorLogger
+      ): Status = {
+        errorCodesVersionSwitcher.choose(
+          v1 = {
+            val statusBuilder =
+              GrpcStatus.toJavaBuilder(Code.ABORTED.value(), Some("Timeout"), Iterable.empty)
+            GrpcStatus.buildStatus(Map.empty, statusBuilder)
+          },
+          v2 = LedgerApiErrors.CommandErrors.CommandCompletionFailure
+            .Reject(
+              message = "Timeout"
+            )
+            .asGrpcStatusFromContext,
+        )
+      }
+
+      def noStatusInResponse()(implicit
+          contextualizedErrorLogger: ContextualizedErrorLogger
+      ): Status = {
+        errorCodesVersionSwitcher.choose(
+          v1 = {
+            val statusBuilder = GrpcStatus.toJavaBuilder(
+              Code.INTERNAL.value(),
+              Some("Missing status in completion response."),
+              Iterable.empty,
+            )
+            GrpcStatus.buildStatus(Map.empty, statusBuilder)
+          },
```

Comment:
Can't we just build a Status directly? Why do we need these two hops?
e.g. 
```suggestion
Can't we just build a Status directly? Why do we need these two hops?
e.g. 
```suggestion
          Status
            .newBuilder()
            .setCode(Code.INTERNAL.value())
            .setMessage("Missing status in completion response.")
            .build(),
```
```

Reply:
Sure!